### ISSUE TITLE
Endian for 256 bit values are still confusing

### ISSUE THREAD

**Issue by joemphilips** [lightningnetwork/lightning-rfc#795]
in https://github.com/lightningnetwork/lightning-rfc/issues/237#issuecomment-327022111, it says LN uses "non-reversed" endian which is the same for bitcoin layer1 wire protocol.
And bolt 01 says, all data fields are big-endian unless specified. And an endian for fields such as `block_hash` are indeed not specified. 
Which implies we are assuming bitcoin layer1 wire protocol is using "big-endian" ,
But this terminology seems to be not universal, e.g. [This Bitcoin Stack Exchange Question](https://bitcoin.stackexchange.com/questions/32765/how-do-i-calculate-the-txid-of-this-raw-transaction)  assumes the value returned from RPC is bigendian, thus byte order for wire protocol (and internal bitcoin-core internal representation) is little-endian.
NBitcoin assumes in that way too (i.e. big-endian for RPC, little-endian for wire)

So I'm confused to use which for my LN implementation.

In short, my question is

1. Is there official document specifying 256bit in wire protocol is indeed "big-endian" ? If that is the case, NBitcoin and the above Bitcoin-SE question is the one which is wrong. If there is not. Than we should clarify about it in the bolt.
2. I think we should define in the bolt that 256-bit values such as "block_hash" should be serialized in the same way that bitcoin does in the layer1 so that big/small endian notation does not matter. Does this make sense?

**Comment by junderw** [comment]
Endianness is a method of converting `bytes[]` to some sort of integer.

block_hash is just `bytes[]` as-is, so endianness is irrelevant.

The only part that is confusing is terminology: Some clients refer to the txid as txhash.

My view on the issue:

blockId and txId = "the big-endian byte representation of the uint256 number checked for PoW" (ie. for blocks "zeroes go on the left")
blockHash and txHash = "the byte array output by the SHA256 x 2 hashing" (ie. for blocks "zeroes go on the right")

Therefore `block_hash` is not ambiguous since it is referring to the hash output, which is a byte array and no integer conversion is needed...

Please correct me if I'm wrong.

**Comment by junderw** [comment]
The reason Bitcoin is so confusing:

```
bytes[] blockHash = doubleSha256(blockHeader);
uint256 blockHashInt = readUInt256AsLittleEndian(blockHash); // MSB is bytes[31] instead of bytes[0]
bytes[] blockId = writeUInt256AsBigEndianBytes(blockHashInt); // MSB is now bytes[0]...... FLIPPED!!!
console.log(blockId.toHex()); // 0000000019db....fec2 "zeroes on the left"
```

But if we just take blockHash and toHex() it we'll get "zeroes on the right" because we don't need the uint256 at all...

**Comment by junderw** [comment]
tl;dr use the hash, not the ID (which is flipped due to the interesting design decision above)

**Comment by joemphilips** [comment]
I see, well some libraries (e.g. NBitcoin which I'm using) does not distinguish those two and uses "little/big" endian notation for all 256-bit value they handle.

So I still think it will help to clarify on the BOLT that `block_hash` has the same byte representation with the bitcoin wire protocol. Because someone who thinks bitcoin wire protocol uses "little-endian" for uint256 might flip the value when reading `chain_hash` fields.

**Comment by knocte** [comment]
> does not distinguish those two and uses "little/big" endian notation for all 256-bit value they handle.

Should we make a PR to NBitcoin too? (Besides clarifying the BOLT)

**Comment by junderw** [comment]
I hate endianness.

It would be incorrect to say "this bytes[] is little endian" because bytes[] do not have endianness except within the context of an integer.

if 0x01020304 === 16909060 the you can say "the bytes[] 0x01020304 is a big-endian representation of 16909060"

But we are talking about a block hash. It is a hash, and only is used as an integer in the context of the PoW check.

So "`block_hash` is a little endian representation of the PoW integer being checked" is a correct statement, but "sha256 outputs a little endian byte array" is incorrect.

**Comment by junderw** [comment]
> Should we make a PR to NBitcoin too? (Besides clarifying the BOLT)

I would be all for it. (aka I believe that NBitcoin is misleading with the naming)

But tbh, I think we should make a PR for Bitcoin :-DDDDDDDDDDDDDD

"Bitcoin Big" BCB a new hard fork coin where all integers are big endian like sane people.

**Comment by junderw** [comment]
If we were to clarify the BOLT I would word it like this:

```
block_hash is the byte array of the little endian representation of the proof of work hash for the block, otherwise known as the byte array output from the double sha256 proof of work hash.
```

**Comment by joemphilips** [comment]
> Should we make a PR to NBitcoin too? (Besides clarifying the BOLT)

It is possible but I bet @NicolasDorier will hesitate to do so. Because introducing the change without breaking the old behavior is quite difficult and awkward.
(I am supposing that the change is to introduce independent type for `TxId`, `WtxId`, `BlockHash` etc... as rust-bitcoin does.)

So I think clarifying in the BOLT is enough. I will make a PR.

**Comment by knocte** [comment]
> But tbh, I think we should make a PR for Bitcoin

Can a PR for bitcoin-core be done but that doesn't change behaviour? (i.e. changes variable names, comments, and docs, to clarify this)
